common/oatbox/reporting/Report.php

```diff
@@ -0,0 +1,441 @@
+ * @method $this containsSuccess() Whenever or not the report contains successes
+ * @method $this containsWarning() Whenever or not the report contains warnings
+ * @method $this containsError() Whenever or not the report contains errors
+ */
+class Report implements IteratorAggregate, JsonSerializable
+{
+    public const TYPE_INFO = 'info';
+
+    public const TYPE_SUCCESS = 'success';
+
+    public const TYPE_WARNING = 'warning';
+
+    public const TYPE_ERROR = 'error';
+
+    /**
+     * Type of the report
+     *
+     * @var string
+     */
+    private $type;
+
+    /**
+     * Message of the report
+     *
+     * @var string
+     */
+    private $message;
+
+    /**
+     * Elements of the report
+     *
+     * @var array
+     */
+    private $children = [];
+
+    /**
+     * Attached to the report data
+     *
+     * @var mixed
+     */
+    private $data;
+
+    /**
+     * Report constructor.
+     *
+     * @param string $type
+     * @param string $message
+     * @param null   $data
+     * @param array  $children
+     *
+     * @throws common_exception_Error
+     */
+    public function __construct(string $type, string $message = '', $data = null, array $children = [])
+    {
+        if (!$this->isValidType($type)) {
+            throw new InvalidArgumentException('Such report type is unsupported');
+        }
+
+        $this->type = $type;
+        $this->message = $message;
+        $this->data = $data;
+
+        foreach ($children as $child) {
+            $this->add($child);
+        }
+    }
+
+    /**
+     * Covers static helpers by the next template: Report::create<Type>
+     *
+     * @param string $name
+     * @param array  $arguments
+     *
+     * @return null|Report
+     * @throws common_exception_Error
+     */
+    public static function __callStatic(string $name, array $arguments)
+    {
+        if (strpos($name, 'create') === 0) {
+            return new static(strtolower(str_replace('create', '', $name)), ...$arguments);
+        }
+
+        return null;
+    }
+
+    /**
+     * Covers helpers (methods) by defined templates
+     *
+     * @param string $name
+     * @param array  $arguments
+     *
+     * @return bool|array|null
+     */
+    public function __call(string $name, array $arguments)
+    {
+        /**
+         * Covers methods by template: get<Type>[e]s
+         */
+        if (strpos($name, 'get') === 0) {
+            $type = strtolower(str_replace('get', '', $name));
+            $type = rtrim(substr($type, 0, -1), 'e');
+            return $this->filterChildrenByTypes([$type], ...$arguments);
+        }
+
+        /**
+         * Covers methods by template: contains<Type>
+         */
+        if (strpos($name, 'contains') === 0) {
+            return $this->contains(strtolower(str_replace('contains', '', $name)));
+        }
+
+        return null;
+    }
+
+    /**
+     * Get message of the report
+     *
+     * @return string
```

Comment:
What additional value `@return` PHPDoc adds on top of strictly defined return type in method signature?

Reply:
Actually, nothing, seems it left after moving code to the new file.

Reply:
WDYT regarding cleanup such leftovers, in the code you've touched ?